library_management_system_downloader/run_api_download_and_parse_results.py

```diff
@@ -0,0 +1,201 @@
+    cursor.close()
+
+# Create an sqlalchemy metadata schema:
+metadata = MetaData(sql_engine)
+
+# Create a schema for the table that will hold the API data:
+# The full_text_indicator column will be an integer in that 1 = fulltext
+# available and 0 = fulltext not available:
+dois_table = Table(
+        'dois_table', metadata,
+        Column('database_id', Integer, primary_key=True),
+        Column('doi', String, nullable=False, unique=True, index=True))
+
+library_holdings_table = Table(
+        'library_holdings_data', metadata,
+        Column('doi_foreign_key',
+               String, ForeignKey("dois_table.database_id"), nullable=False),
+        Column('timestamp', DateTime, nullable=False,
+               default=func.now()),  # Default is creation time
+        Column('xml_response', String, nullable=False),
+        Column('full_text_indicator', Integer))
+
+dois_table.create(
+        checkfirst=True)  # Don't create the table if it already exists.
+
+library_holdings_table.create(
+        checkfirst=True)  # Don't create the table if it already exists.
+
+# Create an sqlalchemy portal for inserting new records into
+# the tables:
+dois_table_insert = dois_table.insert()
+holdings_table_insert = library_holdings_table.insert()
+
+# =============================================================================
+# Define functions for querying and inserting into the database
+# =============================================================================
+
+Session = sessionmaker(bind=sql_engine)
+sql_session = Session()
+
+
+def is_doi_already_answered_in_database(
+        doi,
+        sqlalchemy_session=sql_session):
+    existing_doi_with_fulltext_anwer = sql_session.query(distinct(
+        dois_table.c.doi)).join(library_holdings_table).filter(
+            dois_table.c.doi == doi).filter(
+                    library_holdings_table.c.full_text_indicator.isnot(
+                            None)).first()
+
+    if existing_doi_with_fulltext_anwer is None:
+        return False
+    else:
+        return True
+
+# Example invocation:
+# is_doi_already_answered_in_database('10.1001/2012.jama.10139')
+
+
+def insert_a_doi_database_record(
+        doi,
+        api_response_text,
+        full_text_indicator_for_doi,
+        sqlalchemy_session=sql_session):
+    doi_foreign_key = sql_session.query(
+            distinct(dois_table.c.database_id)).filter(
+                    dois_table.c.doi == doi).first()
+
+    if doi_foreign_key is None:
+        logging.info(f'Inserting DOI "{doi}" into the database...')
+        inserted_doi = dois_table_insert.execute(doi=doi)
+        # Get the foreign key we just inserted into the database:
+        doi_foreign_key = inserted_doi.inserted_primary_key[0]
+    else:
+        doi_foreign_key = doi_foreign_key[0]
+        logging.info(f'DOI "{doi}" is already in the database (foreign key '
+                     '"{doi_foreign_key}"), so not inserting it...')
```

Comment:
The naming here might be unclear -- the DOI itself isn't a foreign key -- rather, `doi_foreign_key` refers to the Primary Key of the `dois` table. The `dois` table currently contains that Primary Key column, which is type `int`, and the DOI as a string. The lookup in this referenced chunk of code is just referencing the `dois` table -- if the DOI is already in that table, it won't get inserted again -- but either way, the download will continue, and will have its output placed into the data table, which is called `library_holdings_data`.

I originally wrote this with a single-table approach, but then came to think that this two-table approach offers more functionality, easier future expansion, and less need for error-catching, for these reasons:

- The main thing that this two-table approach allows is easily re-querying DOIs more than once -- perhaps to see changes over time, or to compare one library with another. With the current code, a configuration setting, `rerun_dois_that_are_already_in_database`, controls whether a DOI that's already been downloaded should be downloaded again. The two-table approach allows this, without requiring re-writing the DOI more than once in a table.
- An earlier draft I wrote did use `DOI` as the Primary Key. But since Primary Keys are unique, that approach wouldn't allow repeat downloads for the same DOI. And having that functionality is both potentially useful for others, and makes it easier for developing (because I can run the same download command several times to quickly check whether the database is working as expected).

Does that make more sense, as I'm explaining it here?

Reply:
> to compare one library with another

If we want to do this, we should add a library column. You could then use (`library`, `doi`, `timestamp`) as the primary key. Note primary keys are not necessary in this single table approach... they just would help us enforce uniqueness constrains, which may or may not be desirable?

> Primary Keys are unique, that approach wouldn't allow repeat downloads for the same DOI

Are there any benefits that wouldn't be solved by placing an index on the DOI column?

Reply:
I don't envision *us* comparing Libraries, but I do see it making sense if others want not only to view the code but to re-use it. So I can go ahead and add a `library` column (or foreign key, as well as a `doi` column, depending on your response to [a comment I just made above](https://github.com/greenelab/library-access/pull/7#discussion_r147198487)).

And I agree that, if DOI isn't serving as Primary Key, placing an index on it would be desirable, and would make the table work fine.